test/rules/avoid_runtimeType_toString.dart

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N avoid_runtimeType_toString`
+
+var o;
+
+class A {
+  var field;
+  String f() {
+    final s1 = '$runtimeType'; // LINT
+    final s2 = runtimeType.toString(); // LINT
+    final s3 = this.runtimeType.toString(); // LINT
+    final s4 = '${runtimeType}'; // LINT
+    final s5 = '${o.runtimeType}'; // OK
+    final s6 = o.runtimeType.toString(); // OK
+    final s7 = runtimeType == runtimeType; // OK
+    final s8 = field?.runtimeType?.toString(); // OK
+    try {
+      final s9 = '${runtimeType}'; // LINT
+    } catch (e) {
+      final s10 = '${runtimeType}'; // OK
+    }
+    throw '${runtimeType}'; // OK
+  }
+}
+
+abstract class B {
+  void f() {
+    final s1 = '$runtimeType'; // OK
+  }
+}
+
+mixin C {
```

Comment:
What should the behavior be inside an extension? I'm guessing it depends on the type being extended.

Reply:
Good question... I made the lint trigger for extension on non-abstract classes.